win32ss/gdi/ntgdi/freetype.c

```diff
@@ -2994,38 +2988,91 @@ static unsigned int get_bezier_glyph_outline(FT_Outline *outline, unsigned int b
+    FT_WinFNT_HeaderRec WinFNT;
+    LONG A, D, AD;
 
-    if (Width < 0)
-        Width = -Width;
+    if (lfWidth < 0)
+        lfWidth = -lfWidth;
 
-    if (Height < 0)
+    if (lfHeight == 0)
     {
-        Height = -Height;
+        lfHeight = dc->ppdev->devinfo.lfDefaultFont.lfHeight;
     }
-    if (Height == 0)
+    if (lfHeight == 0)
     {
-        Height = dc->ppdev->devinfo.lfDefaultFont.lfHeight;
+        lfHeight = lfWidth;
     }
-    if (Height == 0)
+
+    pOS2 = (TT_OS2 *)FT_Get_Sfnt_Table(face, FT_SFNT_OS2);
+    pHori = (TT_HoriHeader *)FT_Get_Sfnt_Table(face, FT_SFNT_HHEA);
+
+    if (!pOS2 || !pHori)
     {
-        Height = Width;
+        error = FT_Get_WinFNT_Header(face, &WinFNT);
+        if (error)
+            return error;
+
+        FontGDI->tmHeight           = WinFNT.pixel_height;
+        FontGDI->tmAscent           = WinFNT.ascent;
+        FontGDI->tmDescent          = FontGDI->tmHeight - FontGDI->tmAscent;
+        FontGDI->tmInternalLeading  = WinFNT.internal_leading;
+        FontGDI->tmEmHeight         = FontGDI->tmAscent - FontGDI->tmInternalLeading;
+        if (FontGDI->tmEmHeight <= 0)
+            FontGDI->tmEmHeight = 1;
+
+        if (lfHeight < 0)
+            lfHeight = -lfHeight;
+        req.type           = FT_SIZE_REQUEST_TYPE_NOMINAL;
+        req.width          = 0;
+        req.height         = (FT_Long)(lfHeight << 6);
+        req.horiResolution = 0;
+        req.vertResolution = 0;
+        return FT_Request_Size(face, &req);
     }
 
-    if (Height < 1)
-        Height = 1;
+    if (lfHeight > 0)
+    {
+        AD = pOS2->usWinAscent + pOS2->usWinDescent;
+        if (AD == 0)
+        {
+            A = pHori->Ascender;
+            D = -pHori->Descender;
+        }
+        else
+        {
+            A = pOS2->usWinAscent;
+            D = pOS2->usWinDescent;
+        }
+        AD = A + D;
+
+        FontGDI->tmAscent = FT_MulDiv(lfHeight, A, AD);
+        FontGDI->tmDescent = FT_MulDiv(lfHeight, D, AD);
+        FontGDI->tmHeight = FontGDI->tmAscent + FontGDI->tmDescent;
+        FontGDI->tmInternalLeading = FontGDI->tmHeight - FT_MulDiv(lfHeight, face->units_per_EM, AD);
+    }
+    else if (lfHeight < 0)
+    {
+        FontGDI->tmAscent = FT_MulDiv(-lfHeight, pOS2->usWinAscent, face->units_per_EM);
+        FontGDI->tmDescent = FT_MulDiv(-lfHeight, pOS2->usWinDescent, face->units_per_EM);
+        FontGDI->tmHeight = FontGDI->tmAscent + FontGDI->tmDescent;
+        FontGDI->tmInternalLeading = FontGDI->tmHeight + lfHeight;
+    }
 
-    if (Width > 0xFFFFU)
-        Width = 0xFFFFU;
-    if (Height > 0xFFFFU)
-        Height = 0xFFFFU;
+    FontGDI->tmEmHeight = FontGDI->tmHeight - FontGDI->tmInternalLeading;
+    if (FontGDI->tmEmHeight <= 0)
+        FontGDI->tmEmHeight = 1;
 
     req.type           = FT_SIZE_REQUEST_TYPE_NOMINAL;
-    req.width          = (FT_Long)(Width << 6);
-    req.height         = (FT_Long)(Height << 6);
+    req.width          = 0;
```

Comment:
Is the width returned as expected by FT_Request_Size? What about the possibility of specifying a custom font width (example for some monospaced fonts...)?